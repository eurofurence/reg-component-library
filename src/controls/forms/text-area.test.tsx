/** @jsxImportSource @emotion/react */

import TextArea from './text-area'
import renderer from 'react-test-renderer'
import { render } from '@testing-library/react'
import userEvent from '@testing-library/user-event'
import { ChangeEventHandler } from 'react'
import * as fc from 'fast-check'

describe('TextArea', () => {
	it('renders correctly', () => {
		fc.assert(fc.property(fc.fullUnicodeString(), fc.fullUnicodeString(), (label, placeholder) => {
			const tree = renderer
				.create(<TextArea name="comments" label={label} placeholder={placeholder}/>)
				.toJSON()

			expect(tree).toMatchSnapshot()
		}), { numRuns: 20, seed: 2562346 })
	})

	it('has a label', () => {
		const { queryByLabelText } = render(
			<TextArea name="comments" label="Comments" placeholder=""/>,
		)

		expect(queryByLabelText('Comments')).toBeTruthy()
	})

	it('enters focus when clicked directly', async () => {
		const { getByRole } = render(
			<TextArea name="comments" label="Comments" placeholder=""/>,
		)

		const textArea = getByRole('textbox')

		expect(textArea).not.toHaveFocus()

		await userEvent.click(textArea)

		expect(textArea).toHaveFocus()
	})

	it('enters focus when clicked through label', async () => {
		const { getByRole, getByText } = render(
			<TextArea name="comments" label="Comments" placeholder=""/>,
		)

		const label = getByText('Comments')
		const textArea = getByRole('textbox')

		expect(textArea).not.toHaveFocus()

		await userEvent.click(label)

		expect(textArea).toHaveFocus()
	})

	it('can handle unicode input', async () => {
		await fc.assert(fc.asyncProperty(fc.fullUnicodeString({ minLength: 1 }), async (input) => {
			const onChange = jest.fn() as jest.MockedFunction<ChangeEventHandler<HTMLTextAreaElement>>
			const { getByRole } = render(
				<TextArea name="comments" label="Comments" placeholder="" onChange={onChange}/>,
			)

			const textArea = getByRole('textbox')

			await userEvent.type(textArea, input)

			expect(onChange.mock.calls[onChange.mock.calls.length - 1]![0].target.value).toEqual(input)
		}))
	}, 30000)

	it('can display unicode', async () => {
		await fc.assert(fc.asyncProperty(fc.fullUnicodeString({ minLength: 1 }), async (input) => {
			const onChange = jest.fn() as jest.MockedFunction<ChangeEventHandler<HTMLTextAreaElement>>
			const { getByRole } = render(
				<TextArea name="comments" label="Comments" placeholder="" value={input} onChange={onChange}/>,
			)

			const textArea = getByRole('textbox')

			await userEvent.type(textArea, input)

			expect(textArea).toHaveValue(input)
		}))
	}, 30000)

	describe('when controlled', () => {
		it('reflects the input value at all times', async () => {
			const onChange = jest.fn() as jest.MockedFunction<ChangeEventHandler<HTMLTextAreaElement>>
			const { getByRole } = render(
				<TextArea name="comments" label="Comments" value="test" placeholder="" onChange={onChange}/>,
			)

			const textArea = getByRole('textbox')

			expect(textArea).toHaveValue('test')

			await userEvent.type(textArea, 'I will be bringing a fursuit.')

			expect(textArea).toHaveValue('test')
		})

		it('raises onChange events when typed in', async () => {
			const onChange = jest.fn() as jest.MockedFunction<ChangeEventHandler<HTMLTextAreaElement>>
			const { getByRole } = render(
				<TextArea name="comments" label="Comments" value="test" placeholder="" onChange={onChange}/>,
			)

			const textArea = getByRole('textbox')

			await userEvent.type(textArea, 'I will be bringing a fursuit.')

			expect(onChange).toHaveBeenCalled()
		})

		it('supports readOnly fields', async () => {
			const { getByRole } = render(
				<TextArea name="comments" label="Comments" value="test" placeholder="" readOnly/>,
			)

			const textArea = getByRole('textbox')

			await userEvent.type(textArea, 'I will be bringing a fursuit.')

			expect(textArea).toHaveValue('test')
		})
	})

	describe('when uncontrolled', () => {
		it('reflects the typed value', async () => {
			const { getByRole } = render(
				<TextArea name="comments" label="Comments" placeholder=""/>,
			)

			const textArea = getByRole('textbox')

			await userEvent.type(textArea, 'I will be bringing a fursuit.')

			expect(textArea).toHaveValue('I will be bringing a fursuit.')
		})

		it('reflects the defaultValue value originally', () => {
			const { getByRole } = render(
				<TextArea name="comments" label="Comments" placeholder="" defaultValue="uwu"/>,
			)

			expect(getByRole('textbox')).toHaveValue('uwu')
		})

		it('raises onChange events when typed in', async () => {
			const onChange = jest.fn() as jest.MockedFunction<ChangeEventHandler<HTMLTextAreaElement>>
			const { getByRole } = render(
				<TextArea name="comments" label="Comments" placeholder="" onChange={onChange}/>,
			)

			const textArea = getByRole('textbox')

			await userEvent.type(textArea, 'I will be bringing a fursuit.')

			expect(onChange).toHaveBeenCalled()
		})
	})
})
