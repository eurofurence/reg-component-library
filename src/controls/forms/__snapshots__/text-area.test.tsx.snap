// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`TextArea renders correctly 1`] = `
<label
  className="css-1k6lc2c"
>
  <div
    className="css-1wds400"
  />
  <textarea
    className="css-w13f35"
    name="comments"
    placeholder="񹇁򆙳"
  />
</label>
`;

exports[`TextArea renders correctly 2`] = `
<label
  className="css-1k6lc2c"
>
  <div
    className="css-1wds400"
  >
    /򫘍𘅮񆐭
  </div>
  <textarea
    className="css-w13f35"
    name="comments"
    placeholder="󲸽r"
  />
</label>
`;

exports[`TextArea renders correctly 3`] = `
<label
  className="css-1k6lc2c"
>
  <div
    className="css-1wds400"
  />
  <textarea
    className="css-w13f35"
    name="comments"
    placeholder=""
  />
</label>
`;

exports[`TextArea renders correctly 4`] = `
<label
  className="css-1k6lc2c"
>
  <div
    className="css-1wds400"
  />
  <textarea
    className="css-w13f35"
    name="comments"
    placeholder="񜩢"
  />
</label>
`;

exports[`TextArea renders correctly 5`] = `
<label
  className="css-1k6lc2c"
>
  <div
    className="css-1wds400"
  >
    􏿼
  </div>
  <textarea
    className="css-w13f35"
    name="comments"
    placeholder="񧽋㋔񍂶򑳲"
  />
</label>
`;

exports[`TextArea renders correctly 6`] = `
<label
  className="css-1k6lc2c"
>
  <div
    className="css-1wds400"
  >
    娱񹮗𛗃𳟁
  </div>
  <textarea
    className="css-w13f35"
    name="comments"
    placeholder="񆠟"
  />
</label>
`;

exports[`TextArea renders correctly 7`] = `
<label
  className="css-1k6lc2c"
>
  <div
    className="css-1wds400"
  />
  <textarea
    className="css-w13f35"
    name="comments"
    placeholder="valueOf"
  />
</label>
`;

exports[`TextArea renders correctly 8`] = `
<label
  className="css-1k6lc2c"
>
  <div
    className="css-1wds400"
  />
  <textarea
    className="css-w13f35"
    name="comments"
    placeholder="l"
  />
</label>
`;

exports[`TextArea renders correctly 9`] = `
<label
  className="css-1k6lc2c"
>
  <div
    className="css-1wds400"
  />
  <textarea
    className="css-w13f35"
    name="comments"
    placeholder="򭮀񯧠򷉄򚣙򜜼󷒛񊂳"
  />
</label>
`;

exports[`TextArea renders correctly 10`] = `
<label
  className="css-1k6lc2c"
>
  <div
    className="css-1wds400"
  >
    "
  </div>
  <textarea
    className="css-w13f35"
    name="comments"
    placeholder="괫󂟥񪽿񷄎񃶟𺸬𩢍"
  />
</label>
`;

exports[`TextArea renders correctly 11`] = `
<label
  className="css-1k6lc2c"
>
  <div
    className="css-1wds400"
  />
  <textarea
    className="css-w13f35"
    name="comments"
    placeholder="𮙊񢛬󯨦򷗸󺳍𚡆𤶮򉸋"
  />
</label>
`;

exports[`TextArea renders correctly 12`] = `
<label
  className="css-1k6lc2c"
>
  <div
    className="css-1wds400"
  >
    __proto__
  </div>
  <textarea
    className="css-w13f35"
    name="comments"
    placeholder="󒸺𩿥񌇉笉󳖷񴥶򡀴󘥄𝹑"
  />
</label>
`;

exports[`TextArea renders correctly 13`] = `
<label
  className="css-1k6lc2c"
>
  <div
    className="css-1wds400"
  >
    񈯾
  </div>
  <textarea
    className="css-w13f35"
    name="comments"
    placeholder="󐶻Ի󢙟򛪰󇍨"
  />
</label>
`;

exports[`TextArea renders correctly 14`] = `
<label
  className="css-1k6lc2c"
>
  <div
    className="css-1wds400"
  >
    argumentsr
  </div>
  <textarea
    className="css-w13f35"
    name="comments"
    placeholder="󤈍􏿺"
  />
</label>
`;

exports[`TextArea renders correctly 15`] = `
<label
  className="css-1k6lc2c"
>
  <div
    className="css-1wds400"
  >
    򭺩󄌄󺨚󁒲󬁀񠩹ㄝ󚛈򃋮󇉠
  </div>
  <textarea
    className="css-w13f35"
    name="comments"
    placeholder="𽏵󿃂􏿺𺃎"
  />
</label>
`;

exports[`TextArea renders correctly 16`] = `
<label
  className="css-1k6lc2c"
>
  <div
    className="css-1wds400"
  >
    ﳽ򔬞󯦱򿯑𷍘򘃵򷥹򏸍󂒏󗰟
  </div>
  <textarea
    className="css-w13f35"
    name="comments"
    placeholder=""
  />
</label>
`;

exports[`TextArea renders correctly 17`] = `
<label
  className="css-1k6lc2c"
>
  <div
    className="css-1wds400"
  >
    $􏿷
  </div>
  <textarea
    className="css-w13f35"
    name="comments"
    placeholder="񋽢򸗭񥕆􂶛󌤋"
  />
</label>
`;

exports[`TextArea renders correctly 18`] = `
<label
  className="css-1k6lc2c"
>
  <div
    className="css-1wds400"
  >
    񣧺򧕾𓞆󷣄𴚏񮋟𙽮򳭆
  </div>
  <textarea
    className="css-w13f35"
    name="comments"
    placeholder="񁸴󌓍'񡩧􉌲򔭋􆓀񺍎񅇡"
  />
</label>
`;

exports[`TextArea renders correctly 19`] = `
<label
  className="css-1k6lc2c"
>
  <div
    className="css-1wds400"
  >
    󫂄񂺛
  </div>
  <textarea
    className="css-w13f35"
    name="comments"
    placeholder="􏿸"
  />
</label>
`;

exports[`TextArea renders correctly 20`] = `
<label
  className="css-1k6lc2c"
>
  <div
    className="css-1wds400"
  >
    󕚱󤐀𔖢󾲳򯀯󒟽򾄢􃄂񂑟򡝴
  </div>
  <textarea
    className="css-w13f35"
    name="comments"
    placeholder="򊞄􌽑򷔫󄱗𸝩􇡫􎶊𯴏"
  />
</label>
`;
